src/probnum/filtsmooth/timeseriesposterior.py

```diff
@@ -0,0 +1,190 @@
+"""Abstract Base Class for posteriors over states after applying filtering/smoothing."""
+
+import abc
+from typing import Optional, Union
+
+import numpy as np
+
+from probnum import _randomvariablelist, random_variables
+from probnum.type import (
+    ArrayLikeGetitemArgType,
+    FloatArgType,
+    RandomStateArgType,
+    ShapeArgType,
+)
+
+DenseOutputLocationArgType = Union[FloatArgType, np.ndarray]
+"""TimeSeriesPosteriors and derived classes can be evaluated at a single location 't'
+or an array of locations."""
+
+DenseOutputValueType = Union[
+    random_variables.RandomVariable, _randomvariablelist._RandomVariableList
+]
+"""Dense evaluation of a TimeSeriesPosterior returns a RandomVariable if evaluated at a single location,
+and a _RandomVariableList if evaluated at an array of locations."""
+
+
+class TimeSeriesPosterior(abc.ABC):
+    """Posterior Distribution over States after time-series algorithms such as
+    filtering/smoothing or solving ODEs.
+
+    Parameters
+    ----------
+    locations :
+        Locations of the posterior states (represented as random variables).
+    states :
+        Posterior random variables.
+    """
+
+    def __init__(self, locations: np.ndarray, states: np.ndarray) -> None:
```

Comment:
Wasn't there a specific array-like type that we always use? I don't completely recall how it was called. But I'm also the least experienced person for the type hints, so np.ndarray might also be fine^^

Reply:
There is numpy.typing.ArrayLike (which I think is what you're referring to), but that type is not compatible with the current numpy version we are using.

Even if we would use the ArrayLike type, I think it would only make sense a line such as `np.asarray(states)` in the first lines (which we may or may not want -- I dont know). If we specify that we want an array, this should be clear either way.